File: Src/BaseSubsystems/RandomUtils.cpp
//---------------------------------------------------------------------------
// RandomUtils.cpp
//---------------------------------------------------------------------------

/**
@file RandomUtils.cpp


@see BaseSubsystems::RandomUtils

@author Alberto Martínez Villarán
@date Abril, 2015
*/


#include "RandomUtils.h"
#include "LUA\ScriptManager.h"

using namespace std;

namespace BaseSubsystems {

	int RandomUtils::_seed = -1;

	int RandomUtils::_id = 0;

	std::default_random_engine RandomUtils::_generator = std::default_random_engine();

	float RandomUtils::Random()
	{
		std::uniform_real_distribution<float> distribution(0, 1);

		//float res = distribution(_generator);
		//_id++;
		//Log::Debug(std::to_string(_id) + " - RND()"+std::to_string(res));
		//return res;

		return distribution(_generator);
	}


	int RandomUtils::RandomRange(int min, int max)
	{
		std::uniform_int_distribution<int> distribution(min, max);

		//int res = distribution(_generator);
		//_id++;
		//Log::Debug(std::to_string(_id) + " - RND("+std::to_string(min)+", "+std::to_string(max)+")"+std::to_string(res));
		//return res;

		return distribution(_generator);
	}

	void RandomUtils::setRandomSeed(int seed)
	{
		_seed = seed;
		_generator.seed(seed);
		//Log::Debug("RandomSeed: "+std::to_string(seed));	
	}

};///BaseSubsystems

File: Exes/media/LUA/RandomUtils.lua
-- RNDID = 0

-- function nextRand()
-- 	-- body
-- 	local id = RNDID + 1
-- 	RNDID = id
-- 	return id
-- end

function RandomBasic()
	local rnd = RandomUtils.Random()
	return rnd
end

function Random(n)
	local rnd = RandomUtils.RandomRange(0,n)
	return rnd

end

function RandomRange(min, max)
	local rnd = RandomUtils.RandomRange(min, max)
	return rnd
end

function RandomElement( tabla )
	local element = RandomUtils.RandomRange(1, table.getn(tabla))
	return element
end


function setRandomSeed(seed)
	RandomUtils.setSeed(seed)
end